keras/models.py

```diff
@@ -64,19 +64,24 @@ def slice_X(X, start=None, stop=None):
 
 def weighted_objective(fn):
     def weighted(y_true, y_pred, weights, mask=None):
-        # it's important that 0 * Inf == 0, not NaN, so I need to mask first
-        masked_y_true = y_true[weights.nonzero()[:-1]]
-        masked_y_pred = y_pred[weights.nonzero()[:-1]]
-        masked_weights = weights[weights.nonzero()]
-        obj_output = fn(masked_y_true, masked_y_pred)
+        '''
+        y_true dimension:  (sample, timestep, dims)
+        y_pred dimension:  (sample, timestep, dims)
+        weights dimension: (sample, timestep, 1)
+        '''
```

Comment:
When a mask is provided, it means that the input has time steps. However the input might not necessarily be `(samples, timesteps, dims)`, it should more generally be considered to be `(samples, timesteps, **)` (e.g. sequences of pictures).